mountpoint-s3/src/inode/readdir.rs

```diff
@@ -0,0 +1,454 @@
+        if let Some(readded) = self.readded.lock().unwrap().take() {
+            return Ok(Some(readded));
+        }
+
+        // Loop because the next entry from the [ReaddirIter] may be hidden from the file system,
+        // either because it has an invalid name or because it's shadowed by another entry.
+        loop {
+            let next = {
+                let mut iter = self.iter.lock().await;
+                iter.next(client).await?
+            };
+
+            if let Some(next) = next {
+                if let Some(lookup) = self.instantiate_remote_inode(next)? {
+                    return Ok(Some(lookup));
+                }
+            } else {
+                return Ok(None);
+            }
+        }
+    }
+
+    /// Re-add an entry to the front of the queue if the consumer wasn't able to use it
+    pub fn readd(&self, entry: LookedUp) {
+        let old = self.readded.lock().unwrap().replace(entry);
+        assert!(old.is_none(), "cannot readd more than one entry");
+    }
+
+    /// Return the inode number of the parent directory of this directory handle
+    pub fn parent(&self) -> InodeNo {
+        self.parent_ino
+    }
+
+    /// Create or update an inode for the given ReaddirEntry.
+    ///
+    /// This returns `Ok(None)` if inode creation fails because the ReaddirEntry should not be
+    /// visible in the file system (e.g. due to shadowing or an invalid inode name).
+    fn instantiate_remote_inode(&self, entry: ReaddirEntry) -> Result<Option<LookedUp>, InodeError> {
+        let (stat, kind) = match &entry {
+            ReaddirEntry::LocalInode { lookup } => return Ok(Some(lookup.clone())),
+            ReaddirEntry::RemotePrefix { .. } => {
+                let stat = InodeStat::for_directory(self.inner.mount_time, Instant::now());
+                (stat, InodeKind::Directory)
+            }
+            ReaddirEntry::RemoteObject { object_info, .. } => {
+                let stat = InodeStat::for_file(
+                    object_info.size as usize,
+                    object_info.last_modified,
+                    Instant::now(),
+                    Some(object_info.etag.clone()),
+                );
+                (stat, InodeKind::File)
+            }
+        };
+        let remote_lookup = RemoteLookup { stat, kind };
+
+        // Short-circuit the update if we know it'll fail because the name is invalid
+        if !valid_inode_name(entry.name()) {
+            warn!(
+                "{} has an invalid name and will be unavailable",
+                entry.shadowed_description()
+            );
+            return Ok(None);
+        }
+
+        let result = self
+            .inner
+            .update_from_remote(self.dir_ino, entry.name(), Some(remote_lookup));
+        match result {
+            Err(InodeError::ShadowedByDirectory(_, _)) => {
+                warn!(
+                    "{} is shadowed by another directory entry with the same name and will be unavailable",
```

Comment:
I am a bit confused here. In the Semantics doc its written Directories will be shadowed by files of the same name.

Reply:
Huh, that documentation is just wrong, we actually do the opposite -- directories shadow files.